Leanplum-SDK/Classes/Features/Actions/LPActionManager.m

```diff
@@ -425,14 +426,59 @@ + (NSString *)messageIdFromUserInfo:(NSDictionary *)userInfo
 
 - (BOOL)isDuplicateNotification:(NSDictionary *)userInfo
 {
-    if ([self.notificationHandled isEqualToString:[LPJSON stringFromJSON:userInfo]] &&
-        [[NSDate date] timeIntervalSinceDate:self.notificationHandledTime] < 10.0) {
-        return YES;
+    BOOL isDuplicate = NO;
+    NSString *messageId;
+    if (userInfo[@"_lpn"]) {
+        messageId = userInfo[@"_lpn"];
+    }
+    if (userInfo[@"_lpm"]) {
+        messageId = userInfo[@"_lpm"];
     }
 
-    self.notificationHandled = [LPJSON stringFromJSON:userInfo];
-    self.notificationHandledTime = [NSDate date];
-    return NO;
+    NSMutableDictionary *handledNotifications = self.handledNotifications;
+
+    [self purgeExpiredHandledNotifications:handledNotifications];
+    if (handledNotifications[messageId]) {
+        isDuplicate = YES;
+    } else {
+        handledNotifications[messageId] = [NSDate date];
+    }
+    self.handledNotifications = handledNotifications;
+    return isDuplicate;
+}
+
+- (NSString *)handledNotificationsKey {
+    return @"lp_action_manager_handled_notifications";
+}
+
+-(NSMutableDictionary *)handledNotifications {
```

Comment:
The plan as per the doc was using VarCache, because it handles encryption and decryption code. Also, does have the same point of responsibility of storing anything caching